tcpdialer.go

```diff
@@ -346,15 +347,47 @@ func (d *TCPDialer) tryDial(
 	ctx, cancelCtx := context.WithDeadline(context.Background(), deadline)
 	defer cancelCtx()
 	conn, err := dialer.DialContext(ctx, network, addr)
-	if err != nil && ctx.Err() == context.DeadlineExceeded {
-		return nil, ErrDialTimeout
+	if err != nil {
+		if ctx.Err() == context.DeadlineExceeded {
+			return nil, wrapDialWithUpstream(ErrDialTimeout, addr)
+		}
+		return nil, wrapDialWithUpstream(err, addr)
 	}
-	return conn, err
+	return conn, nil
 }
 
 // ErrDialTimeout is returned when TCP dialing is timed out.
 var ErrDialTimeout = errors.New("dialing to the given TCP address timed out")
 
+// ErrDialWithUpstream wraps dial error with upstream info.
+//
+// Should use errors.As to get upstream information from error:
+//
+// var dialErr *fasthttp.ErrDialWithUpstream
+//
+//	if errors.As(err, &dialErr) {
+//	  upstream = dialErr.Upstream
+//	}
```

Comment:
```suggestion
//	hc := fasthttp.HostClient{Addr: "foo.com,bar.com"}
//	err := hc.Do(req, res)
//	
//	var dialErr *fasthttp.ErrDialWithUpstream
//	if errors.As(err, &dialErr) {
//		upstream = dialErr.Upstream // foo.com or bar.com
//	}
```
Isn't this a better example of the usage? Isn't this what you build it for?